walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,88 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.PriorityQueue;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    PriorityQueue<TimeRange> busyTimeRanges = getBusyTimeRanges(events, request);
+    busyTimeRanges = combineOverlappingTimeRanges(busyTimeRanges);
+    return getAvailableMeetingTimes(busyTimeRanges, request);
   }
+
+  private PriorityQueue<TimeRange> getBusyTimeRanges(Collection<Event> events,
+      MeetingRequest request) {
+    PriorityQueue<TimeRange> busyTimeRanges = new PriorityQueue<>(TimeRange.ORDER_BY_START);
+
+    // Find all time blocks where at least one meeting attendee is busy (attending another event)
+    for (Event event : events) {
+      if (hasMeetingAttendee(request.getAttendees(), event)) {
+        busyTimeRanges.offer(event.getWhen());
+      }
+    }
+
+    return busyTimeRanges;
+  }
+
+  private boolean hasMeetingAttendee(Collection<String> meetingAttendees, Event event) {
+      for (String eventAttendee : event.getAttendees()) {
+        if (meetingAttendees.contains(eventAttendee)) {
+          return true;
+        }
+      }
+    return false;
+  }
+
+  private PriorityQueue<TimeRange> combineOverlappingTimeRanges(PriorityQueue<TimeRange> timeRanges) {
```

Comment:
I return a PQ here even though I can use another linear data structure (e.g. normal Queue or LinkedList) because it doesn't affect the overall time complexity. The method which takes longest to execute is the getBusyTimeRanges method.